test/e2e.subscribe.geth.js

```diff
@@ -0,0 +1,158 @@
+const assert = require('assert');
+const utils = require('./helpers/test.utils');
+const geth = require('./helpers/test.geth.api');
+const Web3 = utils.getWeb3();
+
+// NB: there are parallel tests for ganache at `eth.subscribe.ganache` in the main unit tests
+describe('subscription connect/reconnect (geth) [ @E2E ]', function() {
+    this.timeout(15000);
+
+    // Geth instamine over WS is unreliable
+    if (!process.env.GETH_AUTOMINE) return;
+
+    let web3;
+    let accounts;
+
+    // Setup a seperate RPC connection for the admin calls
+    before(function(){
+        geth.connectHTTP(Web3, 8545);
+    });
+
+    beforeEach(async function() {
+        web3 = new Web3('ws://localhost:8546');
+        accounts = await web3.eth.getAccounts();
+    });
+
+    afterEach(async function() {
+        await geth.admin.wsStart();
+    });
+
+    // Open connection is given as true by `beforeEach` accounts fetch
+    it('admin.close closes the connection (baseline)', async function(){
+        await geth.admin.wsStop();
+
+        try {
+            const val = await web3.eth.getBlockNumber();
+            console.log('received a val after socket closed: ' + val)
+            assert.fail();
+        } catch (err){
+            console.log('err.message --> ' + err.message);
+            assert(err.message.includes('connection not open on send()'))
+        }
+    });
+
+    it('subscribes (baseline)', function() {
+        return new Promise(async function (resolve) {
+            web3.eth
+                .subscribe('newBlockHeaders')
+                .on('data', function(result) {
+                    assert(result.parentHash);
+                    this.removeAllListeners();
+                    resolve();
+                })
+        });
+    });
+
+    // This test times-out. There *is* a `connection not open on send()` error
+    // received in the callback handler in Subscription.subscribe, but it does not bubble up.
+    it.skip('errors on subscribing when connection is already closed', async function() {
+        await geth.admin.wsStop();
+
+        return new Promise(async function (resolve) {
+            web3.eth
+                .subscribe('newBlockHeaders')
+                .on('error', function(err) {
+                    assert(err.message.includes('CONNECTION ERROR'));
+                    resolve();
+                });
+        });
+    });
+
+    // This test passes on it's own, but not when run in sequence with the others.
+    // Might be easier to debug when test #3 passes
+    it.skip('errors when subscr is already running & connection gets closed', function() {
+        let stage = 0; // Required to not trigger server.close a second time
```

Comment:
I know it is a draft PR but this could get removed with using of the ``once`` method instead of the ``on`` method of the ``EventEmitter`` class the ``Subscription`` class is extending from.

Reply:
Actually I'm seeing things that have no listeners attached fail sometimes, especially in longer test runs. Idk if the problems here are geth, the tests, or the code, or what.... 

I will keep this PR in draft and try it against 1.x for tests 1-5 (which should be already working functionality, maybe). It might be good to have as many working tests on this topic for 1.x before the provider improvements go in so that there's a baseline... we'd know if the tests are bad or should work, whatever. 

Ganache works well but geth seems super race-conditiony. It may be the tests.

Reply:
Great I will do some tests as well! It’s probably worth to think about adding parity to our test suite as well because of the different behaviour we have noticed during writing tests for geth. 👌